Fixed #29522 -- Refactored the Deserializer functions to classes. (#18335)
django/django: docs/topics/serialization.txt

    serializers.deserialize("xml", data, ignorenonexistent=True)

> In addition, you can extend the ``Deserializer`` class and extend it::

Author: I know this may be lame but waiting for your comments. The same for the release note.

Reviewer: I think maybe we should mention how folks can add their own serialization format, so for example we should mention the `SERIALIZATION_MODULES` setting and how they can extend extisting `Deserializer`s

I think this should be within the `Serialization formats` section, maybe at the bottom after `YAML`, as `Custom  serialization formats` (or something). This can have a `.. _custom-serialization-formats:` reference link we can use for the release notes.

It will also need a `.. versionchanged: 5.2` note

Reviewer: I have tried to rewrite the docs a bit, can you review the updates? Would be good to get more eyes on them for sure

Referential fields are again represented by the PK or sequence of PKs.

.. _custom-serialization-formats:

Custom serialization formats
----------------------------

In addition to the default formats, you can create a custom serialization
format.

For example, let’s consider a csv serializer and deserializer. First, define a
``Serializer`` and a ``Deserializer`` class. These can override existing
serialization format classes:

.. code-block:: python
   :caption: ``path/to/custom_csv_serializer.py``

    import csv

    from django.apps import apps
    from django.core import serializers
    from django.core.serializers.base import DeserializationError

>     class Serializer(serializers.python.Serializer):
>         def __init__(self):
>             self.model = None
>             self.content = []
> 
>         def getvalue(self):
>             # Grandparent super
>             return super(serializers.python.Serializer, self).getvalue()
> 
>         def get_dump_object(self, obj):
>             """
>             Serializes a single object into CSV row format.
>             """
>             dumped_object = super().get_dump_object(obj)
> 
>             if self.model is None:
>                 self.model = dumped_object["model"]
>                 self.write_header()
> 
>             row = [dumped_object["model"], str(dumped_object["pk"])]
>             row += [str(value) for value in dumped_object["fields"].values()]
> 
>             return ",".join(row)
> 
>         def write_header(self):
>             fields = [field.name for field in apps.get_model(self.model)._meta.fields]
>             header = ",".join(fields)
>             self.content.append(f"model,{header}\n")
> 
>         def end_object(self, obj):
>             dumped_object_str = self.get_dump_object(obj)
>             self.content.append(f"{dumped_object_str}\n")
> 
>         def end_serialization(self):
>             for line in self.content:
>                 self.stream.write(line)

Reviewer: I haven't tested but I am hoping we can do something like...
```suggestion
    class Serializer(serializers.python.Serializer):
        def get_dump_object(self, obj):
            dumped_object = super().get_dump_object(obj)
            row = [dumped_object["model"], str(dumped_object["pk"])]
            row += [str(value) for value in dumped_object["fields"].values()]
            return ",".join(row), dumped_object["model"]

        def end_object(self, obj):
            dumped_object_str, model = self.get_dump_object(obj)
            if self.first:
                fields = [field.name for field in apps.get_model(model)._meta.fields]
                header = ",".join(fields)
                self.stream.write(f"model,{header}\n")
            self.stream.write(f"{dumped_object_str}\n")

        def getvalue(self):
            return super(serializers.python.Serializer, self).getvalue()
```

    class Deserializer(serializers.python.Deserializer):

>         def deserialize(self):
>             """
>             Deserializes CSV data into model instances.
>             """
>             reader = csv.DictReader(self.stream)
>             model = None
> 
>             for row in reader:
>                 try:
>                     if model is None:
>                         model_name = row.get("model")
>                         model = apps.get_model(model_name)
>                     field_data = {
>                         field.name: row.get(field.name, "") for field in model._meta.fields
>                     }
>                     row["fields"] = field_data
> 
>                     yield from super()._handle_object(row)
>                 except Exception as e:
>                     raise DeserializationError(f"Error deserializing row {row}: {e}")

Reviewer: Ideally we overwrite existing methods (not sure if this works)
```suggestion
        def __init__(self, stream_or_string, **options):
            if isinstance(stream_or_string, bytes):
                stream_or_string = stream_or_string.decode()
            try:
                objects = csv.DictReader(stream_or_string)
            except Exception as exc:
                raise DeserializationError() from exc
            super().__init__(objects, **options)

        def _handle_object(self, obj):
            try:
                model_fields = apps.get_model(obj["model"])._meta.fields
                obj["fields"] = {
                    field.name: obj[field.name]
                    for field in model_fields
                    if field.name in obj
                }
                yield from super()._handle_object(obj)
            except (GeneratorExit, DeserializationError):
                raise
            except Exception as exc:
               raise DeserializationError(f"Error deserializing object: {exc}") from exc
```

Referential fields are again represented by the PK or sequence of PKs.

.. _custom-serialization-formats:

Custom serialization formats
----------------------------

> In addition to the above formats, you can create your own serialization format.

Reviewer 2: ```suggestion
In addition to the default formats, you can create a custom serialization 
format.
```
I tried to rephrase, but feel free to discard it.

First, define a ``Serializer`` and a ``Deserializer`` class. These can override
existing serialization format classes::

    from django.core import serializers

    class Serializer(serializers.python.Serializer): ...

>     class Deserializer(serializers.python.Deserializer): ...

Author: @sarahboyce 
Awesome job! 
IMHO, I don't think custom deserialization quite fits into the  `Custom serialization formats`. Plus, adding a custom serializer to the `SERIALIZATION_MODULES` is a convenient way that only applies to serializers but not such a thing for a custom deserializer. So it doesn't seem cohesive to me.

Reviewer: I think this is how folks are supposed to do it? `get_deserializer` loads the serializers defined by this setting, and then gets the `Deserializer` from it. So this is how you define both a custom serializer and deserializer

Author: Yes, you're right about the usage and how `get_deserializer` returns and I was confused. I wanted to suggest a direct mention of custom deserialization in the section title. Plus, regarding 
`First, define a ``Serializer`` and a ``Deserializer`` class. `
having a custom Deserialization doesn't raise any error in the custom module unless deserialization for that custom format gets called. Maybe it indicates it's necessary (Not a big deal!).

Reviewer: > having a custom Deserialization doesn't raise any error in the custom module unless deserialization for that custom format gets called. Maybe it indicates it's necessary (Not a big deal!).

I'm not quite sure what the suggestion is, do you think this needs a docs warning? Feel free to push up updates :+1:

Author: Got it. I'll wait for other reviewers' comments. If the docs get no comment, it means it's clear enough.

Reviewer 2: I think we should put an example of Serializer and Deserializer with minimal but working code, and not empty.

Reviewer: Yeah I agree, a more realistic example would be beneficial. @AMK9978 do you mind pulling something together?

Author: @pauloxnet @sarahboyce 
I added a minimal functioning csv serializer and deserializer. Although due to the flat nature of csv, it requires a bit more code, I think it is quite useful as an example.

Then add the module containing the serializer definitions to your
:setting:`SERIALIZATION_MODULES` setting::

    SERIALIZATION_MODULES = {

>         "custom": "path.to.custom_serializer",

Reviewer 2: This path here might be confusing if it is not defined above what it is. Maybe we could define above that the code goes in a python file with a certain name and then call it here to help understand that this is the path to the python file (or module) that contains the serializers.

Reviewer: Yes! @AMK9978 something like
```
.. code-block:: python
    :caption: ``path/to/custom_serializer.py``
```
should work

Reviewer 2: Yes, exactly.
